backup_task: remove a component once it is uploaded

tchaikov opened the pull request:
Previously, during backup, SSTable components are preserved in the
snapshot directory even after being uploaded. This leads to redundant
uploads in case of failed backups or restarts, wasting time and
resources (S3 API calls).

This change removes SSTable components from the snapshot directory once
they are successfully uploaded to the target location. This prevents
re-uploading the same files and reduces disk usage.

This change only "Refs" https://github.com/scylladb/scylladb/issues/20655, because, we can further optimize
the backup process, consider:

- Sending HEAD requests to S3 to check for existing files before uploading.
- Implementing support for resuming partially uploaded files.

Fixes https://github.com/scylladb/scylladb/issues/21799
Refs https://github.com/scylladb/scylladb/issues/20655

---

the backup API is not used in production yet, so no need to backport.

mykaul commented:
I don't like sending 'HEAD' - it's yet another request ($$$) to S3.
I'm very much in favor of resumption.

Neither of course are or should be for this PR.

tchaikov commented:
> I don't like sending 'HEAD' - it's yet another request ($$$) to S3. I'm very much in favor of resumption.
> 
> Neither of course are or should be for this PR.

@mykaul i am not sure i am following you. please note, HEAD is for preparing for the resumption. and, the HEAD request is sent only if we resume a previously aborted / failed backup, and we can selectively use it only if a component is relatively large, so we can benefit from resuming the upload. but without HEAD, how can we tell from where we should resume the upload? 

- another option is to store the number of uploaded bytes as a tag to the object, but to read that tag, we need to send yet another request.
- yet another option is to store the number in a system table.

mykaul commented:
> > I don't like sending 'HEAD' - it's yet another request ($$$) to S3. I'm very much in favor of resumption.
> > Neither of course are or should be for this PR.
> 
> @mykaul i am not sure i am following you. please note, HEAD is for preparing for resuming an upload. and, the HEAD request is sent only if we resume a previously aborted / failed backup, and we can selectively use it only if a component is relatively large, so we can benefit from resuming the upload. but without HEAD, how can we tell from where we should resume the upload?

OK - Please see https://rclone.org/s3/#s3-no-head - I thought it was similar.

tchaikov commented:
@mykaul but we don't use HEAD for checking integrity. this is a different story.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14403/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1111/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13876/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2747/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/516/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5865/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14403/)
#### Failed Tests (1/36907):
- [test_ttl_with_load_and_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14403/testReport/junit/alternator_ttl_tests/TestAlternatorTTL/Tests___dtest_with_topology_changes___test_ttl_with_load_and_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_with_load_and_decommission)
#### Build Details:
- Duration: 8 hr 25 min
- Builder: i-0c7cd733c728ef5df (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14422/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13893/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5884/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2764/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14422/)
#### Build Details:
- Duration: 5 hr 20 min
- Builder: spider3.cloudius-systems.com

xemul commented on db/snapshot/backup_task.cc:
```diff
@@ -97,7 +98,12 @@ future<> backup_task_impl::do_backup() {
         //  - http::client::max_connections limitation
-        std::ignore = _client->upload_file(component_name, destination, _progress, &_as).handle_exception([comp = component_name, &ex] (std::exception_ptr e) {
+        std::ignore = _client->upload_file(component_name, destination, _progress, &_as).then([component_name] {
+            // Delete the uploaded component to:
```
It should happen conditionally based on also-to-be-added API parameter

xemul replied:
> named "remove_on_uploaded"

A suggestion -- it can be named "move_files" or something as this is what it effectively going to do

regevran replied:
I am not familiar with such an option today.
To the best of my knowledge rclone `move` files - there is no option to ask it to `copy`.

What is the use case you have in mind?
Keeping the snapshot after uploading it means that now we have two copies of the same SSTables in two places. We'll need to decide which one to use in various scenarios. Why?

xemul commented on db/snapshot/backup_task.cc:
```diff
@@ -97,7 +98,12 @@ future<> backup_task_impl::do_backup() {
+            // 2. Avoid costly S3 existence checks on future backup attempts
+            return remove_file(component_name.native());
+        }).handle_exception([comp = component_name, &ex] (std::exception_ptr e) {
             snap_log.error("Error uploading {}: {}", comp.native(), e);
```
This .handle_exception will collect exceptions from both -- upload_file() and remove_file() -- and it might be hard to tell which part failed exactly

xemul commented on db/snapshot/backup_task.cc:
```diff
@@ -97,7 +98,12 @@ future<> backup_task_impl::do_backup() {
+            // Delete the uploaded component to:
+            // 1. Free up disk space immediately
+            // 2. Avoid costly S3 existence checks on future backup attempts
+            return remove_file(component_name.native());
```
Thinking out loud: what if we ignore remove_file() exception if it arises and process normally? In case of restart due to further error we'll re-upload this file anyway, but in case no more errors happen, we just end up with successful upload and remove the files with subsequent clear-snapshot anyway

tchaikov replied:
makes sense, how about catching the exception raised by `remove_file()`, and note it down with "warning" logging level? because it is not fatal, albeit unexpected.

xemul replied:
Yup, sounds good

regevran replied:
Based on this discussion, isn't it nicer to replace the code, starting at line 88 with:
```cpp
 88         {
 89             auto gh = uploads.hold();
 90             auto component_name = _snapshot_dir / component_ent->name;
 91             auto destination = fmt::format("/{}/{}/{}", _bucket, _prefix, component_ent->name);
 92             snap_log.trace("Upload {} to {}", component_name.native(), destination);
 93 
 94             // Pre-upload break point. For testing abort in actual s3 client usage.
 95             co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
 96 
 97             // Start uploading in the background. 
 98             // The caller waits for these fibers  with the uploads gate.
 99             // Parallelism is implicitly controlled in two ways:
100             //  - s3::client::claim_memory semaphore
101             //  - http::client::max_connections limitation
102             try {
103                 co_await _client->upload_file(component_name, destination, _progress, &_as);
104             } catch (std::exception_ptr e)
105             {
106                 snap_log.error("Error uploading {}: {}", component_name.native(), e);
107                 // keep the first exception
108                 if (!ex) {
109                     ex = std::move(e);
110                 }
111             };
112 
113             try {
114                 co_await remove_file(component_name.native());
115             } catch (std::exception_ptr e) {
116                 // add a warning about the exception
117             }
118         }
```

tchaikov replied:
@regevran i didn't work on weekend. so i didn't get a chance to reply Pavel's comment. i am not sure what you intend to suggest here. but if you want to take this issue, please feel free to reassign it to yourself.

regevran replied:
@tchaikov -
1. No need that you work on weekends.
2. At your work time take a look at the code I _suggested_.
3. I think it is 'nicer', but it is up to you to decide if so, and if I changed it correctly.
4. The main difference is that I am using `co_await` and using the default behavior of a `gate` which is an RAII.

tchaikov commented:
v2:

- do not fail the backup task when fail to remove a component
- added an optional query parameter for "backup" endpoint, so we can optionally move the file instead of copying it
- added a commit to extract `upload_component()` out for better readability.

xemul commented on db/snapshot-ctl.hh:
```diff
@@ -105,7 +105,7 @@ public:
     future<> clear_snapshot(sstring tag, std::vector<sstring> keyspace_names, sstring cf_name);
 
-    future<tasks::task_id> start_backup(sstring endpoint, sstring bucket, sstring prefix, sstring keyspace, sstring table, sstring snapshot_name);
+    future<tasks::task_id> start_backup(sstring endpoint, sstring bucket, sstring prefix, sstring keyspace, sstring table, sstring snapshot_name, bool move_files);
```
Please, use bool_class for the new boolean parameter

tchaikov replied:
actually i was using bool_class in my previous revision, but i dropped it before pushing it to the remote repo. because it does not help with the readability. because `move_files` is always represented with a named variable, instead of literal boolean value, like `true` or `false`. if we use a bool_class, we will have some thing like:

```c++
// shall we add and expose this tag class to the caller as well ? like: 
// auto move_files_ = req_param<bool>(*req, "move_files", false);
// co_await ctl.start_backup(std::move(endpoint), std::move(bucket), std::move(prefix), std::move(keyspace), std::move(table), std::move(snapshot_name), move_files ? move_files_ ? move_files::yes : move_files::no)
// even if we call this function like:
// auto move_files = req_param<bool>(*req, "move_files", false);
// ...
start_backup(std::move(endpoint), std::move(bucket), std::move(prefix), std::move(keyspace), std::move(table), std::move(snapshot_name), move_files);
   auto task = co_await _task_manager_module->make_and_start_task<::db::snapshot::backup_task_impl>(
        {}, *this, std::move(cln), std::move(bucket), std::move(prefix), keyspace, dir, move_files);
   // or
   auto task = co_await _task_manager_module->make_and_start_task<::db::snapshot::backup_task_impl>(
        {}, *this, std::move(cln), std::move(bucket), std::move(prefix), keyspace, dir, move_files);
...
```

not to mention, we have to expose this tag class at a shared place which is visible from both the caller and callee . and one of the callers is in `storage_service.cc` which should not include `backup_task.hh`, as the later is part of the internals of `snapshot-ctl.hh`.

regevran replied:
Can you please explain why are you still dealing with the details when the [requirement](https://github.com/scylladb/scylladb/pull/22285#discussion_r1921515637) for this wasn't answered yet?

tchaikov commented on db/snapshot/backup_task.cc:
```diff
@@ -85,6 +107,7 @@ future<> backup_task_impl::do_backup() {
             break;
         }
         auto gh = uploads.hold();
+        std::ignore = upload_component(component_ent->name).finally([gh = std::move(gh)] {});
```
yes. it was incomplete. should have been fixed in the latest revision.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14613/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14613/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (26/463):
- [test_abort_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.4)
- [test_abort_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.4)
- [test_backup_is_abortable_in_s3_client.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.1)
- [test_backup_is_abortable_in_s3_client.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_1_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.1)
- [test_backup_is_abortable_in_s3_client.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.10)
- [test_backup_is_abortable_in_s3_client.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_10_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.10)
- [test_backup_is_abortable_in_s3_client.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.2)
- [test_backup_is_abortable_in_s3_client.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.2)
- [test_backup_is_abortable_in_s3_client.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.5)
- [test_backup_is_abortable_in_s3_client.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14613/testReport/junit/object_store/test_backup/test_backup_is_abortable_in_s3_client_debug_5_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.5)
#### Build Details:
- Duration: 4 hr 0 min
- Builder: i-010bbd023c42f8d6e (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: UNSTABLE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14629/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14629/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2924/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6044/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14057/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14629/)
#### Build Details:
- Duration: 9 hr 24 min
- Builder: spider1.cloudius-systems.com

regevran reviewed (changes requested):
Please answer this:
https://github.com/scylladb/scylladb/pull/22285#discussion_r1922360273

The default API must be move, with optional to copy - 
then we don't remove the local snapshot.
But the copy option should have a strong basis.

bhalevy commented:
@tchaikov, @xemul doesn't that mean that if we crash in the middle we lost the sstable?
IIUC, if some but not all components were moved, both the local and the remote sstables will miss some components and so should be considered as corrupt.
Why not remove all the local components only after all components were uploaded to remote storage?
This will allow safe retry if we crashed midway, or deduplication if we crashed right after all components were uploaded and the local components weren't deleted.  If we crash while deleting the local components the sstable can be safely removed as well.

bhalevy commented:
https://github.com/scylladb/scylladb/issues/22631
